Serialize/Services.cs

```diff
@@ -1108,7 +1265,8 @@ public static int DecodeAlarmSummary(byte[] buffer, int offset, int apduLen, ref
                 var value = new BacnetAlarmSummaryData();
 
                 len += ASN1.decode_tag_number_and_value(buffer, offset + len, out var tagNumber, out var lenValue);
-                len += ASN1.decode_object_id(buffer, offset + len, out value.objectIdentifier.type, out value.objectIdentifier.instance);
+                len += ASN1.decode_object_id(buffer, offset + len, out BacnetObjectTypes type, out var instance);
+                value.objectIdentifier = new BacnetObjectId(type, instance);
                 len += ASN1.decode_tag_number_and_value(buffer, offset + len, out tagNumber, out lenValue);
                 len += ASN1.decode_enumerated(buffer, offset + len, lenValue, out var tmp);
                 value.alarmState = (BacnetEventStates)tmp;
```

Comment:
I had a long, hard look at the generic implementation of decode_enumerated and decode_context_enumerated... and I think they are evil, and we should probably get rid of them.

since you can't (easily) restrict TEnum to Enum, it currently gets used for uints and object too, which is probably not intentional. Also, since *dynamic* is used, we lose compile-time-type-safety and will only find problems at runtime, when the cast fails.

also, we can only use those generic versions because some classes have public fields that we pass as out param to these functions - but public fields should be avoided (which I tend to forget all the time) so you're back to *out EnumType variableName* and assignment to the **property** on the next line, which ridicules the use of generics in this case.

I tried a few different solutions, but always end up at "use decode_unsigned and cast" :disappointed:

Reply:
Yeah, casting to dynamic and then to target type is a bit of a hack that allows pain free conversion to type that *should* be Enum which we can't really enforce. There is another way by using ```Enum.Parse()``` but that's still error prone if the generic argument is not Enum. Anyway, as you wrote, public fields should be avoided and we will reach that. But the code is relying to heavily on them and usage of out parameters as you already saw. Change this at this stage would not be easy or safe. So for now, I just added the generic version of the method and treat is as the lesser evil, and ditch it once we get rid of public fields and don't rely on out parameters everywhere.